### ISSUE TITLE
[Proposal] Add analyzer to warn about infinite recursion

### ISSUE THREAD

**Issue by jamesqo** [dotnet/roslyn-analyzers#1478]
#### Proposal

We should consider adding an analyzer to warn when a method or property unconditionally recurses (e.g. has no exit points) and will cause a StackOverflowException at runtime. Here's an example of a plausible scenario where this analyzer could help:

```cs
public int X
{
    get => X; // Forgot base.X
    set { AdditionalLogic(); X = value; } // Forgot base.X
}
```

I just ran into a Roslyn bug myself where this would have helped. See https://github.com/dotnet/roslyn/issues/23777

#### Potential False Positives

The analyzer should not issue warnings for the following scenarios:

```cs
void M(int n) {
    if (...) { return ...; } // Possible exit point
    return M(n - 1);
}

void M(int n) {
    if (...) { throw ...; } // Possible exit point
    return M(n - 1);
}
```

#### Other things to consider

- Mutually recursive methods
- Multiple recursion points, e.g.

```cs
// Should fire here
public int M() {
    if (...) {
        return M();
    } else {
        return M();
    }
}
```

**Comment by 333fred** [comment]
Definitely an application for the forthcoming dataflow analysis apis. I don't think a reasonable approximation is good enough here without it.

**Comment by jamesqo** [comment]
@333fred What new data flow analysis APIs are being introduced?

**Comment by 333fred** [comment]
@jamesqo there are no dataflow analysis APIs at all currently. We can't statically determine things like mutual recursion, multiple recursion, or even whether or not recursion conditions can actually be satisfied. We can certainly detect basic conditions like the initial example, but false-positives would be incredibly common for attempting to detect anything beyond the basic example.

**Comment by jamesqo** [comment]
> there are no dataflow analysis APIs at all currently.

What does the `AnalyzeDataFlow` API on `SemanticModel` do then?

**Comment by 333fred** [comment]
`AnalyzeDataFlow` is a relatively limited API that doesn't actually let you define the analysis being performed, it's main purpose is for definite assignment. You couldn't use it to determine if recursion conditions will be met, for example.

**Comment by AlgorithmsAreCool** [comment]
Per some discussion in dotnet/Roslyn#9827 I'll be working on a simplified version of this with no data flow analysis for now that can be enhanced later to cover more scenarios.

**Comment by AlgorithmsAreCool** [comment]
I haven't forgotten about this.
Roslyn is really complicated to know what to do where.

I'll devote some time this weekend to trying again on the implementation.

**Comment by mavasani** [comment]
@jasonmalinowski Another request for the same infinite property assignment issue.

**Comment by jasonmalinowski** [comment]
@mavasani Want to pick one as the primary bug and dupe the rest?

**Comment by mavasani** [comment]
Yes, sounds good.

**Comment by Youssef1313** [comment]
Also requested in #5846

https://github.com/dotnet/roslyn-analyzers/blob/f016189fe1769889f8b247863ef809f12a6c1996/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidInfiniteRecursion.cs could be extended for this case.
